### PR TITLE
add request cancelation, currently only supports cmac

### PR DESCRIPTION
**Pull request by jpbland1**

### CODE REVIEW

**Comment by jpbland1** [comment]
Is this the test bug you guys were talking about or am I actually failing some padding check? regardless, I'm not entirely happy with this solution. It's fine on the server side but I had to complicate the way cancelable requests are handled since cryptoCb can't block and wait for a response the response has to be handled in a new wolfHSM function, `wh_Client_CmacCancelableResponse` which the user has to call after every cmac request, but not if they call init without a key which will confuse the customer. The ideal would be to get everything working with wolfCrypt async since it can already deal with this issue but I'm not sure we should go down that rabbit hole until all the "customer A" requirements have been handled

**Comment by bigbrett** [comment]
> Is this the test bug you guys were talking about or am I actually failing some padding check?
You are actually failling a padding check. One of the CI tests turns on `-Wpadded` which errors out if any padding is applied by the compiler. You added a member to the end of `whClientContext` without adjusting the padding field (also the padding should remain at the end of the struct). 

That said, I think @billphipps slipped this in in a previous PR but I actually think it should be removed as it will be a PITA to maintain. We don't need to manually pad all structs in wolfHSM, we only need to ensure padding on "wire types", meant to be sent across the transport between client/server. I'm gonna add some instrumentation to try and fix this.

File: wolfhsm/wh_transport_mem.h [hunk 1]
```diff
@@ -126,6 +127,7 @@ typedef struct {
     int initialized;
     uint16_t req_size;
     uint16_t resp_size;
+    uint16_t* cancel_seq;
```
Comment (bigbrett) [inline wolfhsm/wh_transport_mem.h:130]: How come this is a pointer, vs just a member of the `whTransportMemContext` struct? AFAICT there isn't a reason it needs this indirection, and the cancellation state could be owned/held by the transport, but could be missing something

File: src/wh_transport_mem.c [hunk 2]
```diff
@@ -205,3 +207,37 @@ int wh_TransportMem_RecvResponse(void* c, uint16_t *out_len, void* data)
...
+    if (*(context->cancel_seq) != 0)
+        return WH_ERROR_NOTREADY;
+
+    /* Check to see if the current response is the different than the request */
+    *(context->cancel_seq) = seq;
+
+    return 0;
+}
+
+int wh_TransportMem_Canceled(void* c, uint16_t* seq)
```
Comment (bigbrett) [inline src/wh_transport_mem.c:229]: same comment as with the callback struct: 

can we rename to something a little more clear and self-describing...perhaps isCanceled or checkCanceledState or something like that? Just something to indicate it is a state query without needing to look at the body of the function and trace the logic through.

File: src/wh_server.c [hunk 3]
```diff
@@ -294,7 +294,13 @@ int wh_Server_HandleRequestMessage(whServerContext* server)
 
         /* Send a response */
         /* TODO: Respond with ErrorResponse if handler returns an error */
-        if (rc == 0) {
+        if (rc == 0 || rc == WH_ERROR_CANCEL) {
+            /* notify the client that their request was canceled */
+            if (rc == WH_ERROR_CANCEL) {
+                kind = WH_MESSAGE_KIND(WH_MESSAGE_GROUP_CANCEL, 0);
```
Comment (bigbrett) [inline src/wh_server.c:300]: Should we make actions for the cancel message group? Even if there is just one for now, probably a good idea

File: src/wh_comm.c [hunk 4]
```diff
@@ -330,6 +331,22 @@ int wh_CommServer_SendResponse(whCommServer* context,
...
+int wh_CommClient_Cancel(whCommClient* context, uint16_t magic)
+{
+    int rc = WH_ERROR_NOTREADY;
+    if (context == NULL || context->transport_cb == NULL ||
+        context->transport_cb->Cancel == NULL) {
+        return WH_ERROR_BADARGS;
+    }
+    context->hdr->magic = magic;
+    context->hdr->kind = wh_Translate16(magic,
+        WH_MESSAGE_KIND(WH_MESSAGE_GROUP_CANCEL, 0));
```
Comment (bigbrett) [inline src/wh_comm.c:343]: Should we make actions for the cancel message group? Even if there is just one for now, probably a good idea.

**Review by bigbrett — COMMENTED** [review]
Nice. 

Still want to do a bit deeper of a dive on this but here are some comments/questions to start.

File: wolfhsm/wh_comm.h [hunk 5]
```diff
@@ -240,6 +249,11 @@ typedef struct {
      */
     int (*Send)(void* context, uint16_t data_size, const void* data);
 
+    /* Check for a canceled request, currently only supports CMAC.
+     * Returns 0 on success, negative number to indicate error.
+     */
+    int (*Canceled)(void* context, uint16_t* seq);
+
```
[the remark this answers is not in the document]
Comment (jpbland1) [reply wolfhsm/wh_comm.h:256]: how about `getCanceledSequence`?

File: wolfhsm/wh_transport_mem.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpbland1) [reply wolfhsm/wh_transport_mem.h:130]: The current cancel callback scheme is transport agnostic, `cancel_seq` is an out-of-band signal for shared memory by virtue of being a separate piece of memory. Implementations that don't use shared memory should set up their own out of band method for passing the current canceled sequence, note that the server calls the callback function directly expecting it's uint16_t pointer to be set but knows nothing with regard to how the implementation achieves this. Same for the client, `wh_CommClient_Cancel` simply calls the Cancel cb with the current sequence but doesn't care how the implementation achieves it. In other words, transport_mem is using shared mem to send the seq but the callers don't see or care how the transport layer sends the signal and other implementations may do it however is necessary.

[the remark this answers is not in the document]
Comment (jpbland1) [reply wolfhsm/wh_transport_mem.h:130]: And I'll add that in the mem transport I simply set cancel_seq to 0 when it is not set to represent no signal but other implementations need to check their own transport and if there's no signal set 0 for the caller manually, might have been part of the confusion.

[the remark this answers is not in the document]
Comment (bigbrett) [reply wolfhsm/wh_transport_mem.h:130]: > The current cancel callback scheme is transport agnostic, cancel_seq is an out-of-band signal for shared memory by virtue of being a separate piece of memory

when I say transport agnostic I mean not requiring modifying the transport layer at all (here you added a bunch of stuff to transport mem, for example).

> Implementations that don't use shared memory should set up their own out of band method for passing the current canceled sequence, note that the server calls the callback function directly expecting it's uint16_t pointer to be set but knows nothing with regard to how the implementation achieves this. Same for the client, wh_CommClient_Cancel simply calls the Cancel cb with the current sequence but doesn't care how the implementation achieves it.

I think the problem is that the state variable _does_ need to be shared between both sides as it is currently written, and that is what I am saying we can't do, since there is no built in way to enforce coherency on that piece of memory (which we would need to do for any region of memory outside the one sequential uncached shared buffer that transport mem uses). 

I'm suggesting that instead of making this a transport specific thing, we make it a port-specific thing entirely done through callbacks. That way you don't need to change anything about the TransportXXX at all. Instead, the client should just invoke a cancel callback registered through the config structure when a "cancel" client function is invoked. And there would be no shared state required. Then on the server side, the cancellation state can be stored locally in the server context, and set out-of-band (e.g. in a server ISR) through a function like `wh_Server_CancelRequest(whServer* server, ...)`. Then any cancellable operation like CMAC can poll the server cancellation state in a similar manner to how it is done now, but through the server context instead of the transport layer. Pretty much entirely mimicking the connection callback functionality. 

tl;dr: I think we should make the cancellation port-specific, not transport-specific,  with no shared state between server and client. This will dramatically simplify things and eliminate any concerns around shared memory access / caching/etc.

**Comment by jpbland1** [comment]
I guess I'm confused as to the layers of the comm client, I thought transport_mem WAS the port specific code, is this not the case? Like for the shared mem unix implementation `wh_transport_mem.c` is the port specific code and has port specific implementations of Init, Recv, etc. from the generic `whTransportServerCb` struct type. `wh_transport_tcp.c` would be the port specific code for the tcp implementation and so on for things like SPI, I2C, etc. when we get to the point. If `wh_transport_mem.c` isn't the port specific code please show me the port code it calls into. As for just using callbacks instead of requiring synchronized variables I think this already does that:
```
int wh_CommClient_Cancel(whCommClient* context, uint16_t magic)
{
    int rc = WH_ERROR_NOTREADY;
    if (context == NULL || context->transport_cb == NULL ||
        context->transport_cb->Cancel == NULL) {
        return WH_ERROR_BADARGS;
    }
    context->hdr->magic = magic;
    context->hdr->kind = wh_Translate16(magic,
        WH_MESSAGE_KIND(WH_MESSAGE_GROUP_CANCEL, 0));
    context->hdr->seq = wh_Translate16(magic, context->seq);
    rc = context->transport_cb->Cancel(context->transport_context,
        context->seq);
    return rc;
}
```
wh_CommClient_Cancel already calls into the generic Cancel callback, which is handled by the port specific `wh_transport_mem.c`. `wh_transport_mem.c` uses shared memory to send the cancel sequence but that's because it's the shared memory port, the TCP port could use out-of-band data, SPI or I2C implementations could use an interrupt and then send the sequence they want to cancel over the transport when it's free next.

File: src/wh_client.c [hunk 6]
```diff
@@ -1087,6 +1093,11 @@ int wh_Client_CmacCancelableResponse(whClientContext* c, Cmac* cmac,
         ret = wh_Client_RecvResponse(c, &group, &action, &dataSz,
             (uint8_t*)packet);
     } while (ret == WH_ERROR_NOTREADY);
+    /* check for out of sequence action */
+    if (ret == 0 && (group != WH_MESSAGE_GROUP_CRYPTO ||
+        action != WC_ALGO_TYPE_CMAC)) {
+        ret = WH_SHE_ERC_SEQUENCE_ERROR;
```
[the remark this answers is not in the document]
Comment (jpbland1) [reply src/wh_client.c]: sorry I missed this comment, changed to abort

File: test/wh_test_crypto.c [hunk 7]
```diff
@@ -653,11 +731,19 @@ int whTest_CryptoServerConfig(whServerConfig* config)
         return WH_ERROR_BADARGS;
     }
 
+    /* set cancelSeqP */
+    cancelSeqP = &server->cancelSeq;
```
Comment (bigbrett) [inline test/wh_test_crypto.c:735]: extremely unlikely, but technically this could be accessed by the client before it is set by the server. Not gonna happen so don't really care, but further case for testing this functionality sequentially, vs in a threaded environment.

**Review by bigbrett — CHANGES_REQUESTED** [review]
Nice, looks great, just a few other things I noticed

File: src/wh_client.c [hunk 8]
```diff
@@ -1019,5 +1082,47 @@ int wh_Client_AesCmacVerify(Cmac* cmac, const byte* check, word32 checkSz,
         ret = memcmp(out, check, outSz) == 0 ? 0 : 1;
     return ret;
 }
+
+int wh_Client_CmacCancelableResponse(whClientContext* c, Cmac* cmac,
```
[the remark this answers is not in the document]
Comment (jpbland1) [reply src/wh_client.c:1086]: No because a response is sent after request, not just the repose to cmac final

File: test/wh_test_crypto.c [hunk 9]
```diff
@@ -653,11 +731,19 @@ int whTest_CryptoServerConfig(whServerConfig* config)
...
+
     WH_TEST_RETURN_ON_FAIL(wh_Server_Init(server, config));
     WH_TEST_RETURN_ON_FAIL(wh_Server_SetConnected(server, am_connected));
     server->comm->client_id = 1;
 
     while(am_connected == WH_COMM_CONNECTED) {
+#ifndef WH_CFG_TEST_NO_CUSTOM_SERVERS
+        while (serverDelay) {
+            sleep(1);
```
[the remark this answers is not in the document]
Comment (jpbland1) [reply test/wh_test_crypto.c:744]: I'll just macro guard it, whole point is to nudge the scheduler into waiting for the client to advance without the server so it'll be smother with a sleep on posix

File: test/wh_test_crypto.c [hunk 10]
```diff
@@ -600,6 +611,73 @@ int whTest_CryptoClientConfig(whClientConfig* config)
...
+#ifndef WH_CFG_TEST_NO_CUSTOM_SERVERS
+        && ret != WH_ERROR_CANCEL_LATE
+#endif
+    ) {
+        WH_ERROR_PRINT("Failed to wh_Client_Cancel %d\n", ret);
+        goto exit;
+    }
+#ifndef WH_CFG_TEST_NO_CUSTOM_SERVERS
+    serverDelay = 0;
+#endif
```
[the remark this answers is not in the document]
Comment (jpbland1) [reply test/wh_test_crypto.c:645]: So here it's `!WH_CFG_TEST_NO_CUSTOM_SERVERS` so if we're using custom servers, don't allow a late cancelation because the `serverDelay` should make sure the cancelation happens in time. `WH_CFG_TEST_NO_CUSTOM_SERVERS`, if we're not using custom server then we can't nudge the scheduler with `serverDelay`, does that make sense?

File: test/wh_test_crypto.c (hunk 7 above)
Comment (jpbland1) [reply test/wh_test_crypto.c:735]: It's fine, this won't happen when custom servers is on due to `serverDelay` and otherwise cancel late will kick in and the test will still pass

**Review by bigbrett — APPROVED** [review]
@jpbland1 Looks great, thanks for working through those issues! Lets definitely follow this up with a PR to add the sequential test like we talked about.
